subsys/net/lib/lwm2m/lwm2m_obj_gateway.h

```diff
@@ -17,4 +19,23 @@
 #define LWM2M_GATEWAY_IOT_DEVICE_OBJECTS_RID 3
 /* clang-format on */
 
+typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);
+
+/**
+ * @brief Register a callback which handles the prefixed messages from the server.
+ *  
+ * The callback gets triggered each time the prefix in front of a recieved lwm2m 
+ * msg path matches the prefix set in the LWM2M_GATEWAY_PREFIX_RID buffer.
+ * 
+ * It must handle the content of the coap message completely. 
+ * In case of success the LwM2M engine will then send the formatted coap message, 
+ * otherwise a coap response code is sent.
+ * @param[in] obj_inst_id Object instance ID generating the callback
+ * @param[in] cb gateway message callback
+ * @return  0 for success, -ENOENT if the message should be handled by the lwm2m
+ *          engine or negative error codes in case of error.
+ */
+void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);
+int lwm2m_gw_handle_req(struct lwm2m_message *msg);
+
```

Comment:
```suggestion
/**
 * @brief A callback which handles the prefixed messages from the server.
 *  
 * The callback gets triggered each time the prefix in front of a recieved lwm2m 
 * msg path matches the prefix set in the LWM2M_GATEWAY_PREFIX_RID buffer.
 * 
 * It must handle the content of the coap message completely. 
 * In case of success the LwM2M engine will then send the formatted coap message, 
 * otherwise a coap response code is sent.
 *
 * Example of returning CoAP response:
 * @code{.c}
 * lwm2m_init_message(msg);
 * // Write CoAP packet to msg->out.out_cpkt
 * return 0;
 * @endcode
 *
 *
 * @return 0 if msg contains a  valid CoAP response.
 * @return  negative error code otherwise. 
 */
typedef int (*lwm2m_engine_gateway_msg_cb)(struct lwm2m_message *msg);

/**
 * @brief Register a callback which handles the prefixed messages from the server.
 *  
 */
void lwm2m_register_gw_callback(uint16_t obj_inst_id, lwm2m_engine_gateway_msg_cb cb);

/**
 * @brief Check if given message is handled by Gateway callback.
 * 
 * @return 0 if msg was handled by Gateawy and contains a valid response. Negative error code otherwise.
 * @return -ENOENT if this msg was not handled by Gateway object.
 */
int lwm2m_gw_handle_req(struct lwm2m_message *msg);

```